samples/bluetooth/broadcast_audio_assistant/src/main.c

```diff
@@ -0,0 +1,469 @@
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);
+
+	bt_conn_unref(broadcast_sink_conn);
+	broadcast_sink_conn = NULL;
+
+	k_sem_give(&sem_sink_disconnected);
+}
+
+static void security_changed_cb(struct bt_conn *conn, bt_security_t level,
+				enum bt_security_err err)
+{
+	if (err == 0) {
+		printk("Security level changed: %u\n", level);
+		k_sem_give(&sem_security_updated);
+	} else {
+		printk("Failed to set security level: %u\n", err);
+	}
+}
+
+static void bap_broadcast_assistant_discover_cb(struct bt_conn *conn, int err,
+						uint8_t recv_state_count)
+{
+	if (err == 0) {
+		printk("BASS discover done with %u recv states\n",
+		recv_state_count);
+		k_sem_give(&sem_bass_discovered);
+	} else {
+		printk("BASS discover failed (%d)\n", err);
+	}
+}
+
+static void bap_broadcast_assistant_add_src_cb(struct bt_conn *conn, int err)
+{
+	if (err == 0) {
+		printk("BASS add source successful\n");
+	} else {
+		printk("BASS add source failed (%d)\n", err);
+	}
+}
+
+static struct bt_bap_broadcast_assistant_cb ba_cbs = {
+	.discover = bap_broadcast_assistant_discover_cb,
+	.add_src = bap_broadcast_assistant_add_src_cb,
+};
+
+static void reset(void)
+{
+	printk("\n\nRestarting...\n\n");
+
+	bt_unpair(BT_ID_DEFAULT, NULL);
+
+	broadcast_sink_conn = NULL;
+	selected_broadcast_id = INVALID_BROADCAST_ID;
+	selected_sid = 0;
+	selected_pa_interval = 0;
+	(void)memset(&selected_addr, 0, sizeof(selected_addr));
+
+	k_sem_reset(&sem_source_discovered);
+	k_sem_reset(&sem_sink_discovered);
+	k_sem_reset(&sem_sink_connected);
+	k_sem_reset(&sem_sink_disconnected);
+	k_sem_reset(&sem_security_updated);
+	k_sem_reset(&sem_bass_discovered);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed_cb
+};
+
+int main(void)
+{
+	int err;
+	struct bt_bap_scan_delegator_subgroup subgroup = { 0 };
+	struct bt_bap_broadcast_assistant_add_src_param param = { 0 };
+
+	err = bt_enable(NULL);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return 0;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	bt_le_scan_cb_register(&scan_callbacks);
+	bt_bap_broadcast_assistant_register_cb(&ba_cbs);
+
+start:
+	scan_for_broadcast_sink();
+
+	err = k_sem_take(&sem_sink_connected, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_sink_connected (err %d)\n", err);
+	}
+
+	err = bt_bap_broadcast_assistant_discover(broadcast_sink_conn);
+	if (err != 0) {
+		printk("Failed to discover BASS on the sink (err %d)\n", err);
+	}
+
+	err = k_sem_take(&sem_security_updated, K_SECONDS(10));
+	if (err != 0) {
+		printk("Failed to take sem_security_updated (err %d), resetting\n", err);
+		bt_conn_disconnect(broadcast_sink_conn, BT_HCI_ERR_AUTH_FAIL);
+
+		if (k_sem_take(&sem_sink_disconnected, K_SECONDS(10)) != 0) {
+			/* This should not happen */
+			return -ETIMEDOUT;
+		}
+
+		reset();
+		goto start;
```

Comment:
Not sure if a label + goto is really warranted here, and in the cases where it is most of the time the jumps should be forward rather than backward. Seems to me like you could just wrap all of this in a `while (true) {` loop, and then the `goto start;` here becomes simply `continue;` (the one on line 466 goes away completely).

Reply:
I agree - oops ;)

Reply:
fixed.

Also, with this last push, I tested having the broadcast audio assistant sample app constantly running, while interchanging different sinks (the broadcast audio sink sample + Samsung Galaxy Buds2 Pro) and different sources (broadcast audio source sample + some nRF52840 Dongles with pre-encoded LC3 music).  Every time, the sinks picked up the sources.

The broadcast_audio_sink sample drops connection to the assistant when the stream starts, but this is an issue outside the scope of this sample and is being looked at separately by @cvinayak @Thalley and us.